# 02 · Baseline: standard GPT, 50M tokens, batch 32

The ordinary residual network, stored activations, plain autograd. Batch 32 x 512 = 16,384 tokens a step,
3,051 steps. Batch 32 is the fixed batch for the first two runs: it is the largest power of two the
**standard** model trains at on this machine (notebook 04 finds the exact ceiling).

In [1]:
import sys, os, json, pathlib, subprocess
IN_COLAB = "google.colab" in sys.modules
if IN_COLAB and not pathlib.Path("ERA_V5").exists():
    subprocess.run(["git", "clone", "--depth", "1", "https://github.com/amukka/ERA_V5.git"], check=True)
    os.chdir("ERA_V5/session13_Distributed_Training")
ROOT = pathlib.Path.cwd()
while not (ROOT / "src").exists() and ROOT != ROOT.parent:
    ROOT = ROOT.parent
os.chdir(ROOT); sys.path.insert(0, str(ROOT))
if not (ROOT / "data" / "train.bin").exists():          # ~5 minutes: stream FineWeb-Edu, train the BPE
    subprocess.run([sys.executable, "tools/prepare_data.py"], check=True)

import torch, platform
from src import experiments as X
from src.train import pick_device
dev = pick_device()
print("device:", dev, "| torch", torch.__version__, "| python", platform.python_version(), "|", platform.platform())
if dev == "mps":
    print(f"MPS recommended working set: {torch.mps.recommended_max_memory()/2**30:.2f} GiB")
if dev == "cuda":
    print(torch.cuda.get_device_name(), f"{torch.cuda.get_device_properties(0).total_memory/2**30:.1f} GiB")
print(json.dumps(json.loads((ROOT / "data" / "meta.json").read_text()), indent=1))

device: mps | torch 2.13.0 | python 3.14.5 | macOS-26.6.2-arm64-arm-64bit-Mach-O
MPS recommended working set: 11.84 GiB
{
 "source": "HuggingFaceFW/fineweb-edu, config sample-10BT, streamed in order",
 "vocab_size": 8192,
 "eos_id": 0,
 "train_tokens": 56000000,
 "train_docs": 46080,
 "val_tokens": 1000000,
 "val_docs": 1024,
 "chars_per_token": 3.886,
 "note": "val.bin comes from the tokenizer's training documents; train.bin from the documents after them",
 "seconds": 314.3
}


**The model** (identical in every run): GPT, 10 layers, d_model 384, 6 heads, MLP 4x, sequence 512,
byte-level BPE vocabulary 8,192 with the head tied to the embedding. **21,053,184 parameters**, of which
17,710,848 are outside the embedding tables. No dropout anywhere, because a reversible backward pass
recomputes each block and a random mask would recompute differently. AdamW (0.9, 0.95, wd 0.1), peak LR
1e-3, 3% warmup then cosine to 10%, gradient clip 1.0, fp32.

In [2]:
base = X.train(name="baseline_standard_b32", variant="standard", batch=X.BASE_BATCH, tokens=X.TOKENS,
               eval_every=250, log_every=10)

{"name": "baseline_standard_b32", "device": "mps", "config": {"vocab_size": 8192, "seq_len": 512, "n_layer": 10, "n_head": 6, "d_model": 384, "variant": "standard", "h": 1.0, "fp_iters": 6, "extra": {}}, "params": 21053184, "params_non_embedding": 17710848, "batch": 32, "tokens_per_step": 16384, "steps": 3051, "peak_lr": 0.001, "warmup": 91, "amp_bf16": 0, "torch": "2.13.0", "machine": "arm64", "processor": "arm"}


step     0/3051  tok   0.02M  loss 9.0926  gnorm 6.631  0 tok/s  0.0 min


memory audit at step 3: exact peak 9,372 MiB in aten._log_softmax_backward_data.default


step   100/3051  tok   1.64M  loss 6.3083  gnorm 0.444  11,028 tok/s  2.5 min


step   200/3051  tok   3.28M  loss 5.9581  gnorm 0.400  11,106 tok/s  5.0 min


            eval @ 4.08M tokens: val loss 5.8774


step   300/3051  tok   4.92M  loss 5.8093  gnorm 0.430  11,080 tok/s  7.5 min


step   400/3051  tok   6.55M  loss 5.5391  gnorm 0.455  11,070 tok/s  9.9 min


            eval @ 8.18M tokens: val loss 5.3962


step   500/3051  tok   8.19M  loss 5.3072  gnorm 0.527  11,239 tok/s  12.4 min


step   600/3051  tok   9.83M  loss 5.0888  gnorm 0.502  11,079 tok/s  14.9 min


step   700/3051  tok  11.47M  loss 5.0089  gnorm 0.475  11,033 tok/s  17.4 min


            eval @ 12.27M tokens: val loss 5.0558


step   800/3051  tok  13.11M  loss 4.8922  gnorm 0.490  11,051 tok/s  19.8 min


step   900/3051  tok  14.75M  loss 4.8350  gnorm 0.501  11,055 tok/s  22.3 min


            eval @ 16.37M tokens: val loss 4.7977


step  1000/3051  tok  16.38M  loss 4.7822  gnorm 0.466  11,292 tok/s  24.8 min


step  1100/3051  tok  18.02M  loss 4.6792  gnorm 0.467  11,041 tok/s  27.2 min


step  1200/3051  tok  19.66M  loss 4.5965  gnorm 0.520  10,979 tok/s  29.7 min


            eval @ 20.46M tokens: val loss 4.5532


step  1300/3051  tok  21.30M  loss 4.5593  gnorm 0.518  11,028 tok/s  32.2 min


step  1400/3051  tok  22.94M  loss 4.4862  gnorm 0.506  11,012 tok/s  34.7 min


            eval @ 24.56M tokens: val loss 4.3768


step  1500/3051  tok  24.58M  loss 4.4184  gnorm 0.520  11,241 tok/s  37.2 min


step  1600/3051  tok  26.21M  loss 4.2903  gnorm 0.517  10,950 tok/s  39.6 min


step  1700/3051  tok  27.85M  loss 4.2762  gnorm 0.545  10,995 tok/s  42.1 min


            eval @ 28.66M tokens: val loss 4.2473


step  1800/3051  tok  29.49M  loss 4.2616  gnorm 0.531  10,991 tok/s  44.6 min


step  1900/3051  tok  31.13M  loss 4.2485  gnorm 0.532  10,992 tok/s  47.1 min


            eval @ 32.75M tokens: val loss 4.1588


step  2000/3051  tok  32.77M  loss 4.1235  gnorm 0.580  11,228 tok/s  49.6 min


step  2100/3051  tok  34.41M  loss 4.1271  gnorm 0.507  10,985 tok/s  52.1 min


step  2200/3051  tok  36.04M  loss 4.0827  gnorm 0.549  10,944 tok/s  54.6 min


            eval @ 36.85M tokens: val loss 4.0948


step  2300/3051  tok  37.68M  loss 4.0824  gnorm 0.536  10,977 tok/s  57.0 min


step  2400/3051  tok  39.32M  loss 4.0070  gnorm 0.510  10,950 tok/s  59.5 min


            eval @ 40.94M tokens: val loss 4.0428


step  2500/3051  tok  40.96M  loss 4.0182  gnorm 0.537  11,197 tok/s  62.1 min


step  2600/3051  tok  42.60M  loss 4.0679  gnorm 0.517  10,966 tok/s  64.6 min


step  2700/3051  tok  44.24M  loss 3.9940  gnorm 0.531  10,975 tok/s  67.1 min


            eval @ 45.04M tokens: val loss 4.0058


step  2800/3051  tok  45.88M  loss 4.0152  gnorm 0.528  10,962 tok/s  69.6 min


step  2900/3051  tok  47.51M  loss 3.9129  gnorm 0.526  10,958 tok/s  72.1 min


            eval @ 49.14M tokens: val loss 3.9847


step  3000/3051  tok  49.15M  loss 4.0283  gnorm 0.535  11,168 tok/s  74.6 min


step  3050/3051  tok  49.97M  loss 3.9708  gnorm 0.529  10,965 tok/s  75.8 min


            eval @ 49.97M tokens: val loss 3.9801


{"tok_per_s": 10982.8, "final_train_loss": 3.9707937240600586, "final_val_loss": 3.980102550983429, "peak_exact_mib": 9371.5, "peak_alloc_mib": 8706.5, "peak_driver_mib": 11835.0, "train_seconds": 4550.0}


In [3]:
keys = ["tokens_trained", "steps", "final_train_loss", "final_train_loss_avg_last5pct", "final_val_loss",
        "tok_per_s", "train_seconds", "peak_exact_mib", "peak_exact_op", "peak_alloc_mib", "peak_driver_mib"]
print(json.dumps({k: base[k] for k in keys}, indent=1))

{
 "tokens_trained": 49971200,
 "steps": 3051,
 "final_train_loss": 3.9707937240600586,
 "final_train_loss_avg_last5pct": 3.9758702754974364,
 "final_val_loss": 3.980102550983429,
 "tok_per_s": 10982.8,
 "train_seconds": 4550.0,
 "peak_exact_mib": 9371.5,
 "peak_exact_op": "aten._log_softmax_backward_data.default",
 "peak_alloc_mib": 8706.5,
 "peak_driver_mib": 11835.0
}
